pkg/kv/kvserver/store.go

```diff
@@ -2166,48 +2162,105 @@ func (s *Store) startRangefeedUpdater(ctx context.Context) {
+			sleepTimer := timeutil.NewTimer()
+			defer sleepTimer.Stop()
+
+			var replIDs roachpb.RangeIDSlice
+			ctx, cancel := s.stopper.WithCancelOnQuiesce(ctx)
+			defer cancel()
 			for {
-				interval := getInterval()
-				if interval > 0 {
-					timer.Reset(interval)
-				} else {
-					// Disable the side-transport.
-					timer.Stop()
-					timer = timeutil.NewTimer()
+				if ctx.Err() != nil {
+					return
 				}
-				select {
-				case <-timer.C:
-					timer.Read = true
-					s.rangefeedReplicas.Lock()
-					replIDs = replIDs[:0]
-					for replID := range s.rangefeedReplicas.m {
-						replIDs = append(replIDs, replID)
+				// Notify each replica with an active rangefeed to check for an
+				// updated closed timestamp.
+				s.rangefeedReplicas.Lock()
+				replIDs = replIDs[:0]
+				for replID := range s.rangefeedReplicas.m {
+					replIDs = append(replIDs, replID)
+				}
+				s.rangefeedReplicas.Unlock()
+				// Sort the range IDs so that we notify in the same order on each
+				// iteration. With the pacing below, this helps to ensure a consistent
+				// closed timestamp lag for each range. The closedts-rangefeed-updater
+				// refresh loop operates independently of incoming closed timestamp
+				// updates, so this does not favor some replicas over others.
+				sort.Sort(&replIDs)
+				// We smear these notifications across the refresh interval to avoid
+				// waking up all the replicas' rangefeed processor and registration
+				// goroutines in a short burst. Doing so without pacing can have a
+				// severe impact on goroutine scheduling latency which impacts
+				// foreground latency, as tens of thousands of goroutines can be
+				// marked as runnable in a short period of time.
+				//
+				// Pacing works as follows:
+				// - We spread closed timestamp notifications to replicas across the
+				//   refresh interval (default 200ms).
+				// - We smear across replicas at the granularity of smearInterval (1ms)
+				//   to avoid putting this goroutine to sleep for excessively short
+				//   periods of time.
+				// - To do so, we compute the number of times this goroutine should
+				//   sleep (smearIntervals).
+				// - We then compute the number of replicas that should be notified
+				//   each time this goroutine runs (smearBatchSize).
+				refreshInterval := getRefreshInterval()
+				if refreshInterval == 0 {
+					select {
+					case <-confCh:
+						// Loop around to use the updated refresh interval.
+						continue
+					case <-ctx.Done():
+						return
 					}
-					s.rangefeedReplicas.Unlock()
-					// Notify each replica with an active rangefeed to check for an updated
-					// closed timestamp.
-					for _, replID := range replIDs {
-						r := s.GetReplicaIfExists(replID)
-						if r == nil {
-							continue
+				}
+				smearIntervals := int((refreshInterval + smearInterval - 1) / smearInterval)
+				smearBatchSize := (len(replIDs) + smearIntervals - 1) / smearIntervals
+				for i, replID := range replIDs {
```

Comment:
I think we want to be more responsive to config changes in this loop. E.g. if by mistake (or intentionally) one puts the `1h` interval, and then wants to reverse it to a reasonably low one, they would be bound to wait for the full interval.